substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -1211,79 +1729,497 @@ public int line() {
+            }
+            if (method != that.method) {
+                return null;
+            }
+            if (line() != that.line()) {
+                return null;
+            }
+            int size = localsSize();
+            if (size != that.localsSize()) {
+                return null;
+            }
+            for (int i = 0; i < size; i++) {
+                NativeImageDebugLocalValueInfo thisLocal = (NativeImageDebugLocalValueInfo) localInfoList.get(i);
+                NativeImageDebugLocalValueInfo thatLocal = (NativeImageDebugLocalValueInfo) that.localInfoList.get(i);
+                if (!thisLocal.equals(thatLocal)) {
+                    return null;
+                }
+            }
+            debugContext.log(DebugContext.DETAILED_LEVEL, "Merge  leaf Location Info : %s depth %d (%d, %d) into (%d, %d)", that.name(), that.depth(), that.lo, that.hi - 1, this.lo, this.hi - 1);
+            // merging just requires updating lo and hi range as everything else is equal
+            this.hi = that.hi;
+
+            return this;
+        }
+
+        public NativeImageDebugLocationInfo split(int stackDecrement, int frameSize) {
+            // this should be for an initial range extending beyond the stack decrement
+            assert lo == 0 && lo < stackDecrement && stackDecrement < hi : "invalid split request";
+            return new NativeImageDebugLocationInfo(this, stackDecrement, frameSize);
+        }
+
+    }
+
+    static final Register[] AARCH64_GPREG = {
+                    AArch64.r0,
+                    AArch64.r1,
+                    AArch64.r2,
+                    AArch64.r3,
+                    AArch64.r4,
+                    AArch64.r5,
+                    AArch64.r6,
+                    AArch64.r7
+    };
+    static final Register[] AARCH64_FREG = {
+                    AArch64.v0,
+                    AArch64.v1,
+                    AArch64.v2,
+                    AArch64.v3,
+                    AArch64.v4,
+                    AArch64.v5,
+                    AArch64.v6,
+                    AArch64.v7
+    };
+    static final Register[] AMD64_GPREG = {
+                    AMD64.rdi,
+                    AMD64.rsi,
+                    AMD64.rdx,
+                    AMD64.rcx,
+                    AMD64.r8,
+                    AMD64.r9
+    };
+    static final Register[] AMD64_FREG = {
+                    AMD64.xmm0,
+                    AMD64.xmm1,
+                    AMD64.xmm2,
+                    AMD64.xmm3,
+                    AMD64.xmm4,
+                    AMD64.xmm5,
+                    AMD64.xmm6,
+                    AMD64.xmm7
+    };
+
+    class ParamLocationProducer {
+        Register[] gpregs;
+        Register[] fregs;
+        int nextGPRegIdx;
+        int nextFPregIdx;
+        int nextStackIdx;
+        int stackParamCount;
+
+        ParamLocationProducer(ResolvedJavaMethod method) {
+            Architecture arch = ConfigurationValues.getTarget().arch;
```

Comment:
Will this work across different OSes? If we have `OSX` or `Windows` instead of Linux?

Reply:
Doh! No, of course, this needs to employ an OS- and CPU-specific call protocol.

I actually wanted class ParamLocationProducer to employ the calling convention configuration data that drives the GraalVM compiler but I had no idea how to access it. So, instead I simply replicated the relevant info per CPU. Is there some simple way to obtain the compiler config as use it as an input to ParamLocationProducer?

For now I will fix this by multiplying cases in the implementation of ParamLocationProducer to respect both CPU and OS. However, it would be clearly be more 'future-proof' if the data that drives ParamLocationProducer was the same as that used by the compiler.